### ISSUE TITLE
Selected slot element ?

### ISSUE THREAD

**Issue by RoXuS** [Morlack/wdio-webcomponents#1]
Hi !

Really thx for your component !!

I have an issue with slotted element.

For example if I have :  

```html
<paper-input>
  <div slot="suffix">
    <paper-icon-button</paper-icon-button>
  </div>
</paper-input>
```
```javascript
const searchButton = 'paper-input div[slot="suffix"] paper-icon-button';
browser.click(searchButton);
```

I can't target this element `An element could not be located on the page using the given search parameters.`.

Have you a solution for this ?

**Comment by RoXuS** [comment]
With this commit : https://github.com/RoXuS/wdio-webcomponents/commit/48c58d02c85838922d8ce9e0a1b2546dcb602d0d

I can use slotted element with a particular selector (for example `:notShadowRoot`) (which indicate to **no** enter in shadowRoot for this element)
Then with this:


```javascript
    const searchButton = 'my-app paper-input:notShadowRoot paper-icon-button[icon="search"]';
    const icon = browser.getAttribute(searchButton, 'icon');
    assert.equal(icon, 'search');
```
It is ok !

But I can't use `browser.click(searchButton);` with this code, I don't know why... If you have an idea... And a better way to ignore shadowRoot for several components...

**Comment by CodeDrivenMitch** [comment]
Hi RoXuS! Thanks for using the lib, I look forward to hearing your experience with it. 

I have explained this particular issue here, in the Readme: https://github.com/Morlack/wdio-webcomponents#shadow-host-inclusion
When querying light DOM children (like you are doing), the element actually does not belong to the custom element. In your example:
- `paper-input` is a custom element with a Shadow Root
- `div[slot=suffix]` is a light DOM child of the paper-input element

This makes it that the `div[slot=suffix]` element does not belong to paper-input, but to my-app. 
So you should use the following code:

```
const searchButton = 'my-app paper-icon-button[icon="search"]';
const icon = browser.getAttribute(searchButton, 'icon');
assert.equal(icon, 'search');
```

I understand this is not ideal, and I myself too have been thinking about a solution. The solutions I had in mind were the following two:

- Write documentation and inform the users of the lib
- Modify the finder code to always query on element further down first. So in your example, the querying would be the following:
  - Query for div[slot=suffix]
  - Query for paper-input
  - Query for paper-icon-button
  - Query for div[slot=suffix]

However, I found that the second solution would sometimes break selectors, and is too implicit. You solution is sound too, but has the implication that when Selenium does get official support for Shadow DOM you will need to rewrite them. Perhaps it could be an option.

What is your preferred solution, and does it work after editing the selector after with this explanation in mind?

**Comment by RoXuS** [comment]
Hi @Morlack,

Ok your solution is great, sorry for my lack of understanding ^^.

Clearly, no good solution... I think that I will keep my solution until selenium integrates the shadow dom. On several component I have a big depth so it's more easily to target the component without change the source code. When selenium will publish a fix I will have to do a simple replace.

Thx again for your component, It saved me a lot of time!!

**Comment by CodeDrivenMitch** [comment]
@RoXuS After using this for an extended period of time I can confirm that the functionality you describe would be very useful. I want to propose a similar solution to what you have.

Instead of using this: `my-app my-custom-element my-custom-sub-element:notShadowRoot .example` 
I want to implement this: `my-app my-custom-element::my-custom-sub-element .example`

When concatenating selectors with `::` the library will execute those selectors in one `querySelector` command, not descending into Shadow DOM. What do you think about this?

I think it looks clearer and the :: implies that the two selectors are run together.

**Comment by RoXuS** [comment]
@Morlack, yeah good idea ! But in your example, it should be : `my-app my-custom-element my-custom-sub-element::.example`, no ?

**Comment by CodeDrivenMitch** [comment]
@RoXuS Ah you are right! Good catch :)
I will implement this in the next few days when I have time

**Comment by RoXuS** [comment]
I can take a look @Morlack (if you are ok) ? Maybe I can update my fork ?

**Comment by CodeDrivenMitch** [comment]
This is included in release 1.0.2. Please try it! 

I have included a wdio testcase using the following selector on the Polymer shop: `shop-app app-header::#tabContainer shop-tabs::.iron-selected::a`

**Comment by RoXuS** [comment]
@Morlack it is ok, thx ^^
